src/cryptonote_basic/connection_context.h

```diff
@@ -66,7 +67,8 @@ namespace cryptonote
     uint16_t m_rpc_port;
     uint32_t m_rpc_credits_per_hash;
     bool m_anchor;
-    //size_t m_score;  TODO: add score calculations
+    int32_t m_score;
+    int m_expect_response;
```

Comment:
Does this limits the protocol now to one request at a time? I would use a set instead.

Reply:
Or even a map/vector so you can send multiple requests of the same type.

Reply:
I don't think we pipeline requests anywhere.

Reply:
Ok, I've thought about this some more. Legit responders should always do in-order responses. This is a side-effect of how the TCP server is designed more than anything. The requesting side does not enforce any order, and can probably issue multiple pipelined requests before a prior response. This is the difference between an `invoke` call and a `notification` call. The chain request and block requests are actually notifications, and the requester doesn't explicitly wait for a response which would block the thread (in the epee design).

Reply:
This system only applies to get objects and get chain, not to things like new txes and new blocks. The former aren't pipelined, but a new tx or block in the middle will not break the check (unless you found a case where it does).

Reply:
With regards to pipelining chain or block requests - a request for block or chain is typically in response to an unsolicited notification message. There's no explicit check to prevent another chain or block request while one is outstanding, but instead appears to be enforced through a combination of checks in `core` and PoW. Not exactly ideal because this potentially makes things more brittle. Provided this is the intent, I don't have any other recommendations at the moment.

Reply:
It might be advisable to at some point explicitly state the design intent here via code comment (which seems to be, perhaps, preventing multiple in-flight requests). From my perspective, disallowing multiple in-flight requests tends to be the correct strategy unless there's a very compelling case for pipelining, as pipelining introduces a lot of subtle complexity relating to resource usage (how many pipelined requests are permitted simultaneously? what state do you need to maintain per-connection? etc). Intuitively, as the core data structure is a list of blocks, a pipeline's utility seems dubious (e.g. requesting successor blocks before validating the current tip could end up wasting tons of cycles trying to compute PoW for bogus successors).